test/functional/wallet_avoidreuse.py

```diff
@@ -240,5 +242,37 @@ def test_fund_send_fund_send(self):
         assert_approx(self.nodes[1].getbalance(), 1, 0.001)
         assert_approx(self.nodes[1].getbalance(avoid_reuse=False), 11, 0.001)
 
+    def test_many_utxos_one_dest(self):
+        '''
+        Test the case where [1] generates a new address A, then
+        [0] sends 1 BTC to A 30 times.
+        [1] spends 11 BTC from A
+        20 of the outputs in A should be used, leaving 10 intact.
+        Those 10 should all be marked used.
+        A single output of 9 btc should be change back to self, and should not be marked used.
+        '''
+        self.log.info("Test many utxos one destination")
+
+        fundaddr = self.nodes[1].getnewaddress()
+        retaddr = self.nodes[0].getnewaddress()
+
+        [self.nodes[0].sendtoaddress(fundaddr, 1) for _ in range(30)]
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # listunspent should show 30 unused 1 btc outputs
+        assert_unspent(self.nodes[1], total_count=30, total_sum=30, reused_supported=True, reused_count=0)
+        # getbalances should show no used, 30 btc trusted
+        assert_balances(self.nodes[1], mine={"used": 0, "trusted": 30})
+
+        self.nodes[1].sendtoaddress(retaddr, 11)
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # listunspent should show 1 unused 9 btc output, and 10 used 1 btc outputs
+        assert_unspent(self.nodes[1], total_count=11, total_sum=19, reused_supported=True, reused_count=10)
+        # getbalances should show 10 used, 9 btc trusted
+        assert_balances(self.nodes[1], mine={"used": 10, "trusted": 9})
```

Comment:
This assert is failing for me with `AssertionError: 0E-8 < [9.999..10.001]`.

Reply:
Thanks for testing -- I have modified the test to work according to the implementation, which has the limitation that it only marks *new* utxos as used when it sees them -- to get the 100% correct behavior the user would have to run a rescan. To be honest, this should probably be fixed..